[all] Sofa add mechanical matrix mapper

olivier-goury opened the pull request:
This PR:
- adds the new component MechanicalMatrixMapper (formerly called MappedMatrixForceField)
- suppress the Cmake option SOFA_WITH_EXPERIMENTAL_FEATURES (Was used for development and is useless now) added in https://github.com/sofa-framework/sofa/pull/276


The component MechanicalMatrixMapper allows to compute the mapped matrices computed by ForceFields under mappings, which are ignored by the SOFA animation loops.
Typical usage would be:
**Node**
MechanicalObject (MO) **q**
MechanicalMatrixMapper (NodeParsed = **MappedNode**) --> computes **J**^T.**K**.**J** and **J**^T.**M**.**J**
.....**MappedNode**
.....MappedMechanicalObject (MMO) **q_m**
.....Mapping (input = ../MO, output = MMO) -> Computes **J**.**q_m**
.....ForceField -> Computes stiffness matrix **K**
.....Mass -> Computes Mass Matrix **M**

Any comments welcome!
(Work with @VannesteFelix @ChristianDuriez )
______________________________________________________
<!--- Please leave this at the end of your message -->
This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings.
- [x] does not generate new unit test failures.
- [x] does not generate new scene test failures.
- [x] does not break API compatibility.
- [x] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

EulalieCoevoet committed:
Revert "Remove unused function "renumberConstraintId""
Corresponding changes are now on the branch removed_renumberConstraintId and should be merged to SOFA using PR process

This reverts commit 3ca5df3cccafae354d9ac487f35c4a46d61887c1.

EulalieCoevoet committed:
Merge remote-tracking branch 'upstream/master'

EulalieCoevoet committed:
Merge remote-tracking branch 'upstream/master'

Commit:
[simulation] ADD MechanicalMatrixMapper

olivier-goury committed:
Remove option SOFA_WITH_ESPERIMENTAL_FEATURES

Commit:
[simulation] Move MechanicalMatrixMapper to module SofaGeneralAnimationLoop

Commit:
[SofaGeneralAnimationLoop] RM useless includes

epernod commented:
I'm sorry, I'm not an expert in this.
Is this new mapper a way to optimise the computation of the mechanical system or is it a new mechanism that was not possible before?

epernod commented on modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.inl:
```diff
@@ -0,0 +1,518 @@
+
+
+template<class DataTypes1, class DataTypes2>
+void MechanicalMatrixMapper<DataTypes1, DataTypes2>::addKToMatrix(const MechanicalParams* mparams,
```
This one is quite a big process. Do you think it could be possible to add a test for this method with known results?

epernod commented on modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.inl:
```diff
@@ -0,0 +1,518 @@
+{
+    sofa::core::behavior::BaseInteractionForceField::init();
+
+    if (mstate1.get() == NULL || mstate2.get() == NULL)
```
you are testing it well here. But there are several access to the mstate1 and 2 in the code. How will the scene behave if mstate are not found?
maybe you could use the paremeter: ComponentState::Invalid;

VannesteFelix replied:
Thanks for the remarks, added in commit https://github.com/sofa-framework/sofa/pull/721/commits/451aefec7860c994b747b667fc97073c3afefc52

Commit:
[SofaGeneralAnimationLoop] Add test to init on component links

guparan commented on modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.cpp:
```diff
@@ -0,0 +1,77 @@
+/******************************************************************************
+*       SOFA, Simulation Open-Framework Architecture, version 1.0 RC 1        *
+*                (c) 2006-2011 MGH, INRIA, USTL, UJF, CNRS                    *
```
2006-2018 ;-)

VannesteFelix replied:
Done with https://github.com/sofa-framework/sofa/pull/721/commits/ac71b0ea494cec0a84deeb537e0566f4e2da2919, thanks for the review!

guparan commented on modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.h:
```diff
@@ -0,0 +1,222 @@
+/******************************************************************************
+*       SOFA, Simulation Open-Framework Architecture, version 1.0 RC 1        *
+*                (c) 2006-2011 MGH, INRIA, USTL, UJF, CNRS                    *
```
2006-2018 ;-)

VannesteFelix replied:
Done

guparan commented on modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.inl:
```diff
@@ -0,0 +1,543 @@
+/******************************************************************************
+*       SOFA, Simulation Open-Framework Architecture, version 1.0 RC 1        *
+*                (c) 2006-2011 MGH, INRIA, USTL, UJF, CNRS                    *
```
2006-2018 ;-)

VannesteFelix replied:
Done

Commit:
[SofaGeneralAnimationLoop] Fix license date

olivier-goury committed:
Add example using MechanicalMatrixMapper

Commit:
[SofaGeneralAnimationLoop] Fix problem at init

Commit:
Merge branch 'Sofa_addMechanicalMatrixMapper' of https://github.com/SofaDefrost/sofa into Sofa_addMechanicalMatrixMapper

olivier-goury commented:
@epernod Thank you for your comments!
When using a direct solver (e.g. SparseLDLSolver), and when the SOFA scene involves a forcefield under a mapping, this component allows to map the stiffness matrix through this mapping (and also the Mass matrix). At the moment SOFA just ignores that.

epernod commented:
[ci-build][with-scene-tests]

epernod commented:
ok but shouldn't be the mapping that automatically does that?

olivier-goury commented:
Hey Erik, 
You're right, this component is a current fix for the issue of mapping matrices through mappings, but ideally it would be an operation called automatically by the animation loop and done by the mapping for example. We will look into that in the future.
Olivier

damienmarchal commented on modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.cpp:
```diff
@@ -0,0 +1,77 @@
+////////////////////////////////////////////    FACTORY    //////////////////////////////////////////////
+SOFA_DECL_CLASS(MechanicalMatrixMapper)
+
+int MechanicalMatrixMapperClass = core::RegisterObject("Partially rigidify a mechanical object using a rigid mapping.")
```
Is this description correct.

olivier-goury replied:
No it's not!

olivier-goury committed:
Update documentation and example location for MechanicalMatrixMapper

olivier-goury commented:
Doxygen doc was added to the component

damienmarchal commented on SofaKernel/modules/SofaBaseMechanics/MechanicalObject.inl:
```diff
@@ -2607,9 +2598,12 @@ void MechanicalObject<DataTypes>::buildIdentityBlocksInJacobian(const sofa::help
+template <class DataTypes>
+void MechanicalObject<DataTypes>::renumberConstraintId(const sofa::helper::vector<unsigned>& /*renumbering*/)
+{
+    this->serr << "MechanicalObject<DataTypes>::renumberConstraintId not implemented in the MatrixDeriv constraint API" << this->sendl;
```
No more serr...and more fundamentally... a pure virtual method is ...pure specifically to force the object inheriting from the parent to not forget to implement them. Providing an implementation like this one sounds like skipping over the safeguard that was put in place. Don't we have a better way to handle those cases ?

guparan replied:
If the override is mandatory, a pure virtual function is the way to go. This compiler-bypassing behavior should not be allowed.

If the override is not mandatory, then non-virtual + msg_warning("You didn't override") seems nice.

damienmarchal replied:
I agree with your Guillaume :)

damienmarchal commented on examples/Components/animationloop/MechanicalMatrixMapperExample.pyscn:
```diff
@@ -0,0 +1,50 @@
+        FEMNode= deformablePartNode.createChild('FEMNode')
+        RigidifiedNode.addChild(FEMNode)
+
+        FEMNode.createObject('Mesh',name='meshInput',src="@../../../../meshDivision/beamMesh")                     
```
You should replace link made by string like "@../../../../meshDivision/beamMesh" with automatically generated link from the target object eg: "@"+meshOfStructure.getPathName() 

It is more robust against change of location & naming of the targeted components.

damienmarchal commented on examples/Components/animationloop/MechanicalMatrixMapperExample.pyscn:
```diff
@@ -0,0 +1,50 @@
+        FEMNode.createObject('MechanicalObject', template='Vec3d',name='beamMecha')
+        FEMNode.createObject('HexahedronFEMForceField', name='HexaFF', src="@meshInput")
+        FEMNode.createObject('RestShapeSpringsForceField', name='restShapeFF', points='@../../../meshDivision/box_roi_fix.indices', stiffness="10000", angularStiffness="10000")
+        FEMNode.createObject('UniformMass', totalmass='0.1')
```
Don't use "0.1" to pass constants you can just write totalmass=1.0

damienmarchal commented on examples/Components/animationloop/MechanicalMatrixMapperExample.pyscn:
```diff
@@ -0,0 +1,50 @@
+        SolverNode= rootNode.createChild('SolverNode')
+        SolverNode.createObject('EulerImplicit',verbose='false')
+        SolverNode.createObject('SparseLDLSolver', name="ldlsolveur")
+        #SolverNode.createObject('MechanicalMatrixMapper', template='Vec3d,Rigid', object1='@./deformablePartNode/beamPart1Mech', object2='@./RigidNode/rigid1', nodeToParse='@./deformablePartNode/FEMNode')
```
Remove useless lines :)

damienmarchal commented on examples/Components/animationloop/MechanicalMatrixMapperExample.pyscn:
```diff
@@ -0,0 +1,50 @@
+import Sofa
+import random	
```
Is it used ?

damienmarchal commented on modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.inl:
```diff
@@ -0,0 +1,542 @@
+    }
+    else
+    {
+        msg_info(this) << "There is no mappedMass";
```
no this.

Commit:
[SofaGeneralAnimationLoop] Clean MechanicalMatrixMapper

olivier-goury committed:
Add doc file next to MechanicalMatrixMapper example.
Update doxygen doc in the header.

olivier-goury commented:
- Doxygen doc has been added in the header file
- Html doc was added in the location of the example in examples/Components/animationLoop

olivier-goury commented:
Hello, 
What is the status for this pull request?  Are you OK to merge it?

guparan commented:
Hi @olivier-goury,
I see there still are some edit requests from @damienmarchal in the comments.
Do you plan to fix them?

damienmarchal commented:
Actually there is no doubt this needs to be edited because it contains files that are not related to the PR.

olivier-goury committed:
Remove renumberConstraintID modifications

damienmarchal commented:
[ci-build][with-scene-tests]

olivier-goury commented:
Hello, do you have more remarks on this PR?
Could we consider going for the merge?
Thanks

damienmarchal commented:
Not for me :)

guparan committed:
[SofaFramework] Delete useless file
